use-package.el

```diff
@@ -1406,6 +1408,67 @@ deferred until the prefix key sequence is pressed."
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;
+;;; :customize
+;;
+
+(defun use-package-normalize/:customize (name-symbol keyword arg)
+  "Normalize use-package customize keyword."
+  (let ((error-msg (format "%s wants a (<symbol> <form> <optional string comment>) or list of these" name-symbol)))
+    (unless (listp arg)
+      (use-package-error error-msg))
+    (dolist (def arg arg)
+      (unless (listp def)
+        (use-package-error error-msg))
+      (cl-destructuring-bind (&optional variable value comment) def
```

Comment:
`cl-destructuring-bind` is a macro defined in `cl-macs.el`, so you will need to `(eval-when-compile (require 'cl-macs))`. I do not, however, think this is the best destructuring option here, as it will throw a non-`use-package`-specific `wrong-number-of-arguments` error if `def` has more elements than it is being destructured into (this also makes the `(> (length def) 3)` check further down redundant).

I suggest using `pcase-let` instead:

```el
(pcase-let ((`(,variable ,value ,comment) def))
  ;; ...
  )
```

This has the added benefit of introducing no further dependencies (apart from the already prevailing Emacs24+ requirement).

Reply:
> `(eval-when-compile (require 'cl-macs))`

FYI, it should rather be `(eval-when-compile (require 'cl-lib))` (e.g., see https://lists.gnu.org/archive/html/emacs-devel/2014-01/msg01003.html).

> I suggest using `pcase-let` instead:

Hmm, I'm not sure that one deals well with a different number of arguments either:

```
(pcase-let BINDINGS &rest BODY)

Like ‘let’ [...]
The macro is expanded and optimized under the assumption that those
patterns *will* match, so a mismatch may go undetected or may cause
any kind of error.
```

Reply:
I always feared that might be the case re: `cl-macs` so thanks for confirming it. :)

My reading of the `pcase` and `pcase-let` documentation is that a backquote-style `QPAT` list pattern will always match, regardless of the length of the list being matched against. In other words, I think the `pcase-let` docstring is warning against uses like `(pcase-let (((and (pred listp) l) "wut")))`, which don't match but which `pcase` would happily skip.

Maybe I'm wrong about this, and maybe destructuring isn't worth it for such simple validation checks.

Reply:
> a backquote-style QPAT list pattern will always match, regardless of the length of the list being matched against.

Well if that were true, how would you distinguish between different list lengths? Perhaps noting that `` `(,one ,two ,three)`` is a shorthand for `` `(,one ,two ,three . nil)`` might make it clearer:

```elisp
(mapcar (lambda (x)
          (pcase x
            (`(,one ,two ,three . nil) (list :exactly-three-args one two three))
            (`(,one ,two ,three . ,rest) (list :more-than-three-args one two three rest))
            (_ (list :less-than-three-args x))))
        '((1) (1 2) (1 2 3) (1 2 3 4)))
;=>
((:less-than-three-args
  (1))
 (:less-than-three-args
  (1 2))
 (:exactly-three-args 1 2 3)
 (:more-than-three-args 1 2 3
                        (4)))
```

Reply:
Going with

``` elisp
(let ((var (car def))
      (val (cadr def))
      (com (caadr def)) ...)
```
then ?

Reply:
You mean `(caddr def)`, but otherwise seems fine.

Reply:
indeed, that proves what's nice about destructuring bind ;)

Reply:
@npostavs I see, thanks for clarifying.

@canatella I don't think Emacs 24 has `caddr` yet. Why not just use `nth` for all of them? That way you afford a tiny bit more destructional clarity.